File: Old Code/WurmRancher/WurmRancher/WeedLevel.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;

namespace WurmRancher
{
    public class WeedLevel : TimedLevel
    {
        public WeedLevel(Theme theme) : base(theme, 70) { }

        public override int WeedGrowthRate
        {
            get
            {
                return Timing.RelativeTimeToFrames(2);
            }
        }

        public override string Name
        {
            get
            {
                return "Weed Invasion!";
            }
        }
        public override string Description
        {
            get
            {
                return string.Format("This area seems to be particularly overgrown. No matter-- just use your spray!  Grow your wurm to length {0} to win!  And while your at it, make sure the weed ratio is less than {1}.",length_to_win,ratio_to_win);
            }
        }

        public override string QuickObjectives
        {
            get
            {
                return string.Format("Grow your wurm to length {0} and get the weed ratio down to {1:#.000}.", length_to_win, ratio_to_win);
            }
        }

        const int length_to_win = 10;
        const double ratio_to_win = .20;
        Wurm theWurm;
        Counter WeedDensityCounter;

        public override void InitializeLevel(GameControl theControl)
        {
            base.InitializeLevel(theControl);
            theWurm = new Wurm(theControl, 3, new Point(100, 100));
            theWurm.LengthChange += new EventHandler<GameEventArgs>(theWurm_Grows);
            theControl.AddCounter(new WurmCounter(theWurm));
            WeedDensityCounter = new Counter("Weed Density");
            theControl.AddCounter(WeedDensityCounter);
            while (theControl.WeedRatio < .4)
                theControl.GrowRandomWeed();
            WeedDensityCounter.Value = string.Format("{0:#.000}", theControl.WeedRatio);
            
        }

        void theWurm_Grows(object sender, GameEventArgs e)
        {
            if (theWurm.Length >= length_to_win && e.theControl.WeedRatio <= ratio_to_win)
                Victory(e.theControl);
        }

        public override void Update(GameControl theControl)
        {
            base.Update(theControl);
            if (GameOver)
                return;
            if (IntervalTimeIsUp(1))
                theControl.GrowRandomWeed();
            if (IntervalTimeIsUp(6))
                theControl.AddCreature(new GrassEater(theControl));
            if (IntervalTimeIsUp(3))
                theControl.AddCreature(new Feeder(theControl));
            if (IntervalTimeIsUp(15))
                theControl.AddCreature(new Monster(theControl));

            if (IntervalTimeIsUp(1))
            {
                double weed_ratio = theControl.WeedRatio;
                WeedDensityCounter.Value = string.Format("{0:#.000}", weed_ratio);
                if (theWurm.Length >= length_to_win && weed_ratio <= ratio_to_win)
                    Victory(theControl);
            }

        }

    }
}

File: ts/weedLevel.ts
import { Counter } from "./counter.js";
import { Feeder } from "./feeder.js";
import { AddCounter, WeedRatio, GrowRandomWeed, AddCreature, AddCreatureOnEdge } from "./gameControl.js";
import { GrassEater } from "./grassEater.js";
import { Monster } from "./monster.js";
import { Theme } from "./theme.js";
import { TimedLevel } from "./timedLevel.js";
import { Wurm } from "./wurm.js";
import { WurmCounter } from "./wurmCounter.js";

export class WeedLevel extends TimedLevel{
    WeedGrowthRate = 2000;
    Name = "Weed Invasion!";
    length_to_win = 10;
    ratio_to_win = .20;

    theWurm : Wurm;
    WeedDensityCounter : Counter;

    constructor(theme:Theme){
        super(theme, 70000);
        this.Description = `This area seems to be particularly overgrown. No matter-- just use your spray!  Grow your wurm to length ${this.length_to_win} to win!  And while your at it, make sure the weed ratio is less than ${this.ratio_to_win}.`;
        this.QuickObjectives = `Grow your wurm to length ${this.length_to_win} and get the weed ratio down to ${this.ratio_to_win.toFixed(2)}.`;
    }

    InitializeLevel(): void {
        super.InitializeLevel();
        this.theWurm = new Wurm(3, 100,100);
        AddCounter(new WurmCounter(this.theWurm));
        this.WeedDensityCounter = new Counter("Weed Density");
        AddCounter(this.WeedDensityCounter);
        while (WeedRatio() < .4)
            GrowRandomWeed();
        this.WeedDensityCounter.Value = WeedRatio().toFixed(2);
    }

    Update(time_step: number): void {
        super.Update(time_step);

        if (this.IntervalTimeIsUp(1000))
            GrowRandomWeed();
        if (this.IntervalTimeIsUp(6000))
            AddCreatureOnEdge(new GrassEater());
        if (this.IntervalTimeIsUp(3000))
            AddCreatureOnEdge(new Feeder());
        if (this.IntervalTimeIsUp(15000))
            AddCreatureOnEdge(new Monster());

        if ((this.theWurm.Length >= this.length_to_win) && this.IntervalTimeIsUp(500))
        {
            let weed_ratio = WeedRatio();
            this.WeedDensityCounter.Value =  WeedRatio().toFixed(2);
            if (weed_ratio <= this.ratio_to_win)
                this.Victory();
        }
    }
}